docs/enhancements/infrastructure-inventory-services-api/resource-server.md

```diff
@@ -0,0 +1,372 @@
+- Updates the `generation_id` of the data source being synchronized
+- Deletes entries in the main tables that have a `generation_id` less than that of the data source's `generation_id`
+- Adds a change event to the outbox table for any inserted, updated, or deleted tuples in the main tables
+- Sends a signal event to the subscription processor for each new tuple in the outbox table
+- Receives signal events back from the subscription processor when outbox tuples have been consumed and then deletes the
+  outbox tuples
+- Listens for asynchronous notifications from data sources (if supported) and updates the main tables (and outbox table)
+
+## Tooling and general dev guidelines
+
+- All the features stated above will be developed within in the same server.
+- The HTTP server should be built with latest Go 1.22 `net/http` std lib. The latest update in the package brings in
+  many requested features including mapping URI pattern. This allows us to not rely on third party lib such as
+  `gorilla/mux`.
+- Prefer creating structs to hold HTTP data for idiomatic Go code.
+- OpenAPI spec should be the source of truth. Other than standardization, free validation and documentation,
+  with it, we can leverage a code generator such [this](https://github.com/oapi-codegen/oapi-codegen), allowing us to
+  avoid writing boilerplate code.
+- For Postgres communication use library [pgx](https://github.com/jackc/pgx) v5. This Go Postgres driver and lots of
+  important features such as automatic type mapping, detailed error reporting (capture performance info) etc.
+  There's also many ORM and SQL query builder libraries but pgx looks like the best of both worlds.
+- DB migration is generally handled with a different tool. [golang-migrate](https://github.com/golang-migrate/migrate)
+  is generally used for this which we can call during service init.
+- Introduce server specific commands. In this case we need something that can init a new DB and migrate as needed E.g
+  ```shell
+  oran-o2ims resource-server db-migration -h
+  ```
+
+## Interactions with other services
+
+### Alarm Server
+
+The system is divided into multiple microservices each with their own area of responsibility. Each microservice
+maintains its data in a logical database instances hosted in a shared database service. In some cases there is a need
+for one microservice to access another microservice's data. Rather than directly couple those services by sharing
+access to each other's logical database schema this implementation will add API endpoints between microservices to
+push changes and to query data via API endpoints. Specifically,
+
+- The resource server will push any changes to the resource type table to the alarm server
+- The resource server will pull the latest alarm dictionary data on startup
+- The alarm server will push any changes to the alarm dictionary table to the resource server
+- The alarm server will pull the latest resource type data on startup.
```

Comment:
Cool basically we will fill in the details for each other. Alarms will ask something like `/ResourceType GET`...and then we generate the AlarmDictionary rows and respond back with ResourceType and its corresponding AlarmDictionary filled out. We can probably reuse the exact the O-RAN payload for req and response to keep it simple.

Reply:
yes, we can probably use the 'standard' API endpoints... maybe.